xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,721 @@
+              .setNode(node)
+              .addResourceNames(resourceName)
+              .setTypeUrl(ADS_TYPE_URL_RDS)
+              .setResponseNonce(rdsRespNonce)
+              .build();
+      requestWriter.onNext(request);
+    }
+
+    /**
+     * Prepares a DiscoveryRequest containing the given information as an ACK.
+     */
+    private void prepareAckRequest(String typeUrl, String resourceName, String versionInfo,
+        String nonce) {
+      pendingAckRequest =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(versionInfo)
+              .setNode(node)
+              .addResourceNames(resourceName)
+              .setTypeUrl(typeUrl)
+              .setResponseNonce(nonce)
+              .build();
+    }
+
+    /**
+     * Sends the previously prepared ACK DiscoveryRequest to management server. Updates the latest
+     * accepted version for the corresponding resource type.
+     */
+    private void sendPendingAckRequest() {
+      checkState(pendingAckRequest != null, "There is no pending ACK request");
+      checkState(requestWriter != null, "ADS stream has not been started");
+      // Update the latest accepted version.
+      String typeUrl = pendingAckRequest.getTypeUrl();
+      if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+        ldsVersion = pendingAckRequest.getVersionInfo();
+      } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+        rdsVersion = pendingAckRequest.getVersionInfo();
+      }
+      // TODO: cases for (VHDS/)CDS/EDS.
+      requestWriter.onNext(pendingAckRequest);
+      pendingAckRequest = null;
+    }
+
+    /**
+     * Cancels the pending ACK request and sends its corresponding (with version reverted)
+     * NACK request.
+     */
+    private void nackPendingAckRequest() {
+      checkState(pendingAckRequest != null, "There is no pending ACK request");
+      checkState(requestWriter != null, "ADS stream has not been started");
+      DiscoveryRequest.Builder nackRequestBuilder = pendingAckRequest.toBuilder();
+      // Reverts the request version to the latest accepted version.
+      String typeUrl = pendingAckRequest.getTypeUrl();
+      if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+        nackRequestBuilder.setVersionInfo(ldsVersion);
+      } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+        nackRequestBuilder.setVersionInfo(rdsVersion);
+      }
+      // TODO: cases for (VHDS/)CDS/EDS.
+      requestWriter.onNext(nackRequestBuilder.build());
+      pendingAckRequest = null;
+    }
+  }
+
+  /**
+   * Returns {@code true} iff {@code hostName} matches the domain name {@code pattern} with
+   * case-insensitive.
+   *
+   * <p>Wildcard pattern rules:
+   * <ol>
+   * <li>A single asterisk (*) matches any domain.</li>
+   * <li>Asterisk (*) is only permitted in the left-most or the right-most part of the pattern,
+   *     but not both.</li>
+   * </ol>
+   */
+  @VisibleForTesting
+  static boolean matchHostName(String hostName, String pattern) {
```

Comment:
I see the code from `OkHostnameVerifier.verifyHostName()` being used here (also used in `SdsX509TrustManager.verifyDnsNameInPattern'). At some point we should consider creating a common utility function that can be used by all 3 sites.

Reply:
They have slightly different specification. Again, this is internal helper method, I don't see the necessity of sharing code with other modules.